### ISSUE TITLE
Crash with method call removal optimization

### ISSUE THREAD

**Issue by rasky** [golang/go#14771]
Please answer these questions before submitting your issue. Thanks!

1. What version of Go are you using (`go version`)?
go version devel +53900ce Thu Mar 10 17:46:46 2016 +0000 darwin/amd64

2. What operating system and processor architecture are you using (`go env`)?
darwin/amd64

3. What did you do?
After compiling an existing codebase (https://github.com/rasky/ndsemu) with latest go tip, I get a crash on a method value call, probably related to the recent optimization that removes method calls

4. What did you expect to see?
The program should start

5. What did you see instead?
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x0]

goroutine 1 [running, locked to thread]:
panic(0x4283ea0, 0xc8200100f0)
	/Users/rasky/Sources/gosrc/src/runtime/panic.go:500 +0x189
reflect.callMethod(0xc8200ab290, 0xc82004f768)
	/Users/rasky/Sources/gosrc/src/reflect/value.go:630 +0x1ad
reflect.methodValueCall(0x0, 0x4087541, 0xc8200a6228, 0xffffff0000000000, 0x0, 0x40230aa, 0x445f508, 0x4d81ee0, 0x0, 0x0, ...)
	/Users/rasky/Sources/gosrc/src/reflect/asm_amd64.s:29 +0x36
ndsemu/emu/hwio.(*Reg32).write(0xc8200a6228, 0xffffff0000000000)
	/Users/rasky/Sources/go/src/ndsemu/emu/hwio/reg.go:172 +0x55
ndsemu/emu/hwio.(*Reg32).Write8(0xc8200a6228, 0x4000208)
	/Users/rasky/Sources/go/src/ndsemu/emu/hwio/reg.go:236 +0x2f1
ndsemu/emu/hwio.(*Table).Write8(0xc8205af500, 0x4000208)
	/Users/rasky/Sources/go/src/ndsemu/emu/hwio/table.go:278 +0x453
ndsemu/arm.(*Cpu).opWrite8(0xc8205b6480, 0x4000208)
	/Users/rasky/Sources/go/src/ndsemu/arm/mem.go:172 +0x154
ndsemu/arm.(*Cpu).opArm5C0(0xc8205b6480, 0xe5c44208)
	/Users/rasky/Sources/go/src/ndsemu/arm/ops_arm_table.go:14126 +0x79
ndsemu/arm.(*Cpu).Run(0xc8205b6480, 0x648)
	/Users/rasky/Sources/go/src/ndsemu/arm/ops.go:248 +0x346
main.(*NDS7).Run(0xc8205c6000, 0x648)
	/Users/rasky/Sources/go/src/ndsemu/nds7.go:105 +0x2e
ndsemu/emu.syncSubsystem.Run(0x4d81c00, 0xc8205c6000, 0x100, 0x42e878b, 0x4, 0x648)
	/Users/rasky/Sources/go/src/ndsemu/emu/sync.go:85 +0x5b
ndsemu/emu.(*Sync).RunUntil(0xc8205c2780, 0x648)
	/Users/rasky/Sources/go/src/ndsemu/emu/sync.go:383 +0x17c
ndsemu/emu.(*Sync).RunOneFrame(0xc8205c2780)
	/Users/rasky/Sources/go/src/ndsemu/emu/sync.go:342 +0xad
main.(*NDSEmulator).RunOneFrame(0xc82009c1e0, 0xc820960000, 0x100, 0x1da, 0x400)
	/Users/rasky/Sources/go/src/ndsemu/emulator.go:262 +0x1dd
main.main()
	/Users/rasky/Sources/go/src/ndsemu/ndsemu.go:250 +0x8f6
```

I think the culprit is in `emu/hwio/table.go`, around line 232:

```
			if meth := val.Addr().MethodByName(rcb); !meth.IsValid() {
				return fmt.Errorf("cannot find method: %q", rcb)
			} else {
				valueField.FieldByName("ReadCb").Set(meth)
			}
```

where valueField comes from a structure like:

```
type Reg32 struct {
	Name   string
	Value  uint32
	RoMask uint32

	Flags   RegFlags
	ReadCb  func(val uint32) uint32
	WriteCb func(old uint32, val uint32)
}
```

It looks like the method is now removed from the code generation.

**Comment by rasky** [comment]
@crawshaw

**Comment by martisch** [comment]
Thanks for the report.
The problem with MethodByName is known and @crawshaw is working on a fix, see: #14740 .

**Comment by minux** [comment]
Merged into #14740.
